Sally7/ValueConversion/BufferHelper.cs

```diff
@@ -0,0 +1,69 @@
+﻿using Sally7.Internal;
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Sally7.ValueConversion;
+
+internal class BufferHelper
+{
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static int CopyAndAlign64Bit(ReadOnlySpan<byte> input, Span<byte> output, int numberOfItems)
+    {
+        ref var destination = ref MemoryMarshal.GetReference(output);
+        ref var source = ref MemoryMarshal.GetReference(input);
+
+        var limit = (uint) numberOfItems * sizeof(ulong);
+
+        var offset = 0u;
+        while (offset < limit)
```

Comment:
This assumes that after the loop no remaining elements in the input-span are present. This is OK for the use case here (?), but would you mind adding an assert for this? E.g.
```c#
Debug.Assert(offset == input.Length, "input span's length is not a multiple of sizeof(ulong)");
```
Then the code is self documenting.

Reply:
I'm afraid I need your advice on this. I cheated the input of these methods by doing `Unsafe.As<long[], byte[]>(ref value)`. As a result, for an array of 8 `long` values, I get a `byte[]` with it's `Length` set to... 8. Then in `CopyAndAlign64Bit` I take the pointer to that buffer and just copy the actual number of bytes needed.

So the question is, did I mess up? I could use `MemoryMarshal.Cast<long, byte>(value)`, but if you look at the [benchmarks](https://github.com/mycroes/Benchmarks/commit/ebe032a118c1c738ea02094ab679fb69c1930ccd), there's quite some difference. But using `Unsafe.As<,>()`  means I'm messing with the length, and as such I lose the validation. As a workaround, I could pass in the original array and apply the conversions in the method, but still I won't have `input.Length` that matches the actual length, so the only approach I can take is to calculate based on `numberOfItems`, which is the same calculation as `limit` is ergo I'm validating if I can do the same calculation twice...

Let me know what you think. I think I currently favor the existing approach that relies on me writing the code correctly, but for sure I can't do that 100% of the time.

Reply:
You're right, and maybe I was a bit too conservative here -- all in all it's an internal type and we are under full controll what's passed in. So (now) I think just a little comment like
```c#
// The data passed in is guaranteed to be powers of two long, so no remaining elements will exists.
```
or similar.

In general and where possible I prefer the `Debug.Assert` as it docuemts the code, and that assert can be checked in CI by a debug build too. But as you said, it's not easy to pass in the required info w/o writing more code than needed.